lib/perl/Genome/Model/Tools/Speedseq/AlignBase.pm

```diff
@@ -0,0 +1,141 @@
+
+use strict;
+use warnings;
+
+use Genome;
+
+class Genome::Model::Tools::Speedseq::AlignBase {
+    is => 'Genome::Model::Tools::Speedseq::Base',
+    is_abstract => 1,
+    has_input => [
+        reference_fasta => {
+            is => 'Text',
+            doc => 'fasta file (indexed with bwa)',
+            example_values => ['reference.fa'],
+        },
+    ],
+    has_param => [
+        output_prefix => {
+            is => 'Text',
+            doc => 'output prefix',
+            is_optional => 1,
+            example_values => ['in1.fq'],
+        },
+        paired => {
+            is => 'Boolean',
+            doc => 'first fastq file consists of interleaved paired-end sequences',
+            is_optional => 1,
+        },
+        threads => {
+            is => 'Integer',
+            doc => 'threads',
+            default_value => 1,
+            is_optional => 1,
+        },
+        temp_directory => {
+            is => 'Text',
+            doc => 'temp directory',
+            example_values => ['./output_prefix.XXXXXXXXXXXX'],
+            is_optional => 1,
+        },
+        insert_size => {
+            is => 'Text',
+            doc => 'specify the mean, standard deviation (10% of the mean if absent), max(4 sigma from the mean if absent) and min of the insert size distribution.  FR orientation only. Inferred by default.',
+            is_optional => 1,
+            example_values => ['FLOAT[,FLOAT[,INT[,INT]]]'],
+        },
+        include_duplicates =>{
+            is => 'Boolean',
+            doc => 'include duplicates in splitters and discordants',
+            is_optional => 1,
+        },
+        maximum_split_alignments => {
+            is => 'Integer',
+            doc => 'maximum number of split alignments for a read to be included in splitter file',
+            is_optional => 1,
+            example_values => ['2'],
+        },
+        minimum_non_overlapping_bp => {
+            is => 'Integer',
+            doc => 'minimum non-overlapping base pairs between two alignments for a read to be included in splitter file',
+            is_optional => 1,
+            example_values => ['20'],
+        },
+        sort_memory => {
+            is => 'Integer',
+            doc => 'amount of memory in GB to be used for sorting',
+            is_optional => 1,
+            example_values => ['20'],
+        },
+    ],
+    has_optional_output => [
+        output_files => {
+            is_many => 1,
+        },
+    ],
+};
+
+sub _output_extensions {
+    my $class = shift;
+    return qw/
+                 .bam
+                 .bam.bai
+                 .discordants.bam
+                 .discordants.bam.bai
+                 .splitters.bam
+                 .splitters.bam.bai
+             /;
+}
+
+sub _resolve_options_string {
+    my $self = shift;
+    my $params = shift;
+
+    unless ($params) {
+        $params = '';
+    }
+    if ($self->output_prefix) {
```

Comment:
I would set a parameter map hash like:
%hash = (o => "output_prefix", t => "threads" ...);
then loop through the hash to set parameter option (also considering boolean or not) if defined that parameter.